warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,449 @@
+      }
+
+      conn.setDoInput(true);
+      conn.setRequestMethod(method.toUpperCase());
+
+      //
+      // Set body
+      //
+
+      if (body instanceof String) {
+        String bodyS = (String) body;
+        conn.setDoOutput(bodyS.length() > 0);
+        if (bodyS.length() > 0) {
+          try (OutputStream os = conn.getOutputStream()) {
+            os.write(bodyS.getBytes(StandardCharsets.UTF_8));
+          }
+        }
+
+      } else if (body instanceof byte[]) {
+        byte[] bodyB = (byte[]) body;
+        conn.setDoOutput(bodyB.length > 0);
+        if (bodyB.length > 0) {
+          try (OutputStream os = conn.getOutputStream()) {
+            os.write(bodyB);
+          }
+        }
+
+      } else if (null != body) {
+        throw new WarpScriptException(getName() + " expects the body of the request to be a STRING or BYTES object.");
+      }
+
+      //
+      // Form response
+      //
+
+      res.put(STATUS_CODE, conn.getResponseCode());
+      Map<String, List<String>> hdrs = conn.getHeaderFields();
+
+      if (hdrs.containsKey(null)) {
+        List<String> statusMsg = hdrs.get(null);
+        if (statusMsg.size() > 0) {
+          res.put(STATUS_MESSAGE, statusMsg.get(0));
+        } else {
+          res.put(STATUS_MESSAGE, "");
+        }
+      } else {
+        res.put(STATUS_MESSAGE, "");
+      }
+
+      //
+      // Make the headers map modifiable
+      //
+
+      hdrs = new LinkedHashMap<String, List<String>>(hdrs);
+      hdrs.remove(null);
+
+      res.put(RESPONSE_HEADERS, hdrs);
+
+      //
+      // Read response
+      //
+
+      InputStream in = null;
+
+      // When there is an error (response code is 404 for instance), body is in the error stream.
+      try {
+        in = conn.getInputStream();
+      } catch (IOException ioe) {
+        in = conn.getErrorStream();
+      }
+
+      if (chunkSize <= 0) {
+
+        if (null != in) {
+          byte[] buf = new byte[8192];
+          ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+          while (true) {
+            int len = in.read(buf);
+            if (len < 0) {
+              break;
+            }
+
+            if (downloadSize.get() + baos.size() + len > maxsize) {
+              throw new WarpScriptException(getName() + " would exceed maximum size of content which can be retrieved via this function (" + maxsize + " bytes)");
+            }
+
+            baos.write(buf, 0, len);
+          }
+
+          downloadSize.addAndGet(baos.size());
+          res.put(CONTENT, baos.toByteArray());
+
+        } else {
+          res.put(CONTENT, new byte[0]);
+        }
+
+      } else {
+
+        if (null != in) {
+
+          int chunkNumber = 0;
+          while (true) {
+            chunkNumber++;
+
+            Map<String, Object> chunkRes = new LinkedHashMap<>(res);
+            byte[] buf = new byte[chunkSize.intValue()];
+            int len = in.read(buf);
+            if (len < 0) {
```

Comment:
If EOF is reached while attempting to read buf.length bytes you will exit the loop and miss the last part of the data